colorizer_data/utils.py

```diff
@@ -302,3 +304,158 @@ def copy_remote_or_local_file(src_path: str, dst_path: str) -> None:
+
+
+def cast_feature_to_info_type(
+    data: np.ndarray, info: FeatureInfo
+) -> Tuple[np.ndarray, FeatureInfo]:
+    """
+    Validates the type of a feature, casting the data values if needed.
+    If the feature info has no type (`FeatureType.INDETERMINATE`), attempts to infer the feature's type
+    and updates the feature info and data array.
+
+    - For `FeatureType.DISCRETE` features, float values will be truncated and cast to int (if possible).
+    - For `FeatureType.CONTINUOUS` features, values will be cast to float.
+    - For `FeatureType.CATEGORICAL` features, data values converted to integer indexes into the
+    `info.categories` array. If category information is missing or data is non-numeric, categories
+    will be automatically inferred.
+
+    Args:
+        data (np.ndarray): The feature's data array.
+        info (FeatureInfo): The feature's metadata.
+
+    Raises:
+        RuntimeError if the feature type is CONTINUOUS or DISCRETE, but data is non-numeric.
+
+    Returns:
+        A tuple, containing an `np.ndarray` and a (possibly updated) copy of `info`.
+    """
+    info = info.clone()
+
+    if info.type == FeatureType.INDETERMINATE:
+        logging.warning(
+            "Info type for feature '{}' is INDETERMINATE. Will attempt to infer feature type.".format(
+                info.get_name()
+            )
+        )
+        info.type = infer_feature_type(data, info)
+
+    kind = data.dtype.kind
+    if info.type == FeatureType.CONTINUOUS:
+        if kind not in {"f", "u", "i"}:
+            raise RuntimeError(
+                "Feature '{}' has type set to CONTINUOUS, but has non-numeric data.".format(
+                    info.get_name()
+                )
+            )
+        return (data.astype(float), info)
+    if info.type == FeatureType.DISCRETE:
+        if kind not in {"f", "u", "i"}:
+            raise RuntimeError(
+                "Feature '{}' has type set to DISCRETE, but has non-numeric data.".format(
+                    info.get_name()
+                )
+            )
+        return (safely_cast_array_to_int(data), info)
+    if info.type == FeatureType.CATEGORICAL:
+        if info.categories is not None and kind in {"i", "u", "f"}:
+            # Formatted correctly, return directly
+            return (safely_cast_array_to_int(data), info)
+        # Attempt to parse the data
+        if info.categories == None:
+            logging.warning(
+                "Feature '{}' has type set to CATEGORICAL, but is missing a categories array.".format(
+                    info.get_name()
+                )
+            )
+            logging.warning("Categories will be automatically inferred from the data.")
+            logging.warning(
+                "If the output looks incorrect, provide the categories as a string array and the data as an array of integers."
+            )
+        else:
+            # Feature has predefined categories. Warn that values will be remapped.
```

Comment:
Oof, I was secretly hoping no one would ask for this because it makes the logic for parsing categories a lot harder 😂

Thank you for the feedback, I'll make an attempt at this!

Reply:
Implemented!